File: android/crypto/presentation/src/test/java/dev/ohoussein/crypto/presentation/reducer/CryptoListReducerTest.kt
package dev.ohoussein.crypto.presentation.reducer

import app.cash.turbine.test
import dev.ohoussein.crypto.presentation.model.Crypto
import dev.ohoussein.cryptoapp.common.resource.DataStatus
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import org.mockito.kotlin.mock

class CryptoListReducerTest : BehaviorSpec({

    given("a reducer") {
        val reducer = CryptoListReducer()

        `when`("a UpdateCryptoList event") {
            val newCryptoList: List<Crypto> = mock()
            val event = CryptoListEvents.UpdateCryptoList(
                cryptoList = newCryptoList,
            )
            reducer.sendEvent(event)

            then("it should set the event's data") {
                reducer.state.test {
                    awaitItem() shouldBe CryptoListState(
                        cryptoList = newCryptoList,
                        status = DataStatus.Success,
                    )
                }
            }
        }

        `when`("a UpdateStatus event") {
            val status: DataStatus = mock()
            val event = CryptoListEvents.UpdateStatus(
                status = status,
            )
            reducer.sendEvent(event)

            then("it should set the event's status") {
                reducer.state.test {
                    awaitItem().status shouldBe status
                }
            }
        }
    }
})

File: CryptoAppiOS/CryptoAppiOSTests/Crypto/Reducer/CryptoListReducerTest.swift
@testable import CryptoAppiOS
import XCTest

final class CryptoListReducerTest: XCTestCase {
    func testUpdateStatus() throws {
        let initialState = CryptoListState()

        let newState = cryptoListReducer(state: initialState, event: .updateStatus(.loading))

        XCTAssertEqual(LoadingStatus.loading, newState.status)
    }

    func testUpdateCryptoList() throws {
        let initialState = CryptoListState()
        let cryptoList = CryptoDataMock.mockedCryptoList()

        let newState = cryptoListReducer(state: initialState, event: .updateCryptoList(cryptoList))

        XCTAssertEqual(cryptoList, newState.cryptoList)
    }
}
